File: Binary Search/74.search-a-2-d-matrix.cpp
/*
 * @lc app=leetcode id=74 lang=cpp
 * [74] Search a 2D Matrix
 */
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        //Error3: special case [[]] 
        if(matrix.empty() || matrix[0].empty()) return false;
        int row = matrix.size(); //hang
        int column = matrix[0].size();  //lie

        //Error1: l & r are both index!
        //int l = matrix[0][0], r=matrix[row-1][0]; //[l,r]
        int l = 0, r = row-1;
        while(l<=r){ 
            int mid = (l+r)/2;
            //Tip: [l,r] -> [l,mid-1];[mid+1,r];
            if(matrix[mid][0]==target) return true;
            else if(matrix[mid][0]<target) l=mid+1;
            else r=mid-1;
        }
        //Error4: check if r<0
        if(r<0) return false;
        //Error2:return right not right -1
        int rowIndex = r;
        
        l = 0;
        r = column-1;
        while(l<=r){
            int mid = (l+r)/2;
            if(matrix[rowIndex][mid]==target) return true;
            else if(matrix[rowIndex][mid]<target) l=mid+1;
            else r=mid-1;
        }
        return false;
    }
};

File: Binary Search/74.search-a-2-d-matrix.java
/*
 * @lc app=leetcode id=74 lang=java
 *
 * [74] Search a 2D Matrix
 */

// @lc code=start
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m=matrix.length;
        if(m==0) return false;
        int n=matrix[0].length;
        if(n==0) return false;
        int l=0, r = m*n,mid=0;
        while(l<r){
            mid = (l+r)/2;
            int cx = mid/n, cy = mid%n;
            if(matrix[cx][cy]==target) return true;
            else if(matrix[cx][cy]<target) l=mid+1;
            else r=mid;
        }
        return false;
    }
    
}
// @lc code=end

